src/main/java/org/bytedeco/javacpp/tools/Parser.java

```diff
@@ -1031,6 +1056,9 @@ else if (cppNameSplit.size() > 1 && groupNameSplit.size() == 1)
             }
             type.javaName = context.shorten(type.javaName);
         }
+        type.explicitUpcast =
+            info != null && info.explicitUpcast
+                 || Templates.isSmartPointer(type.cppName) && type.arguments != null && type.arguments[0].explicitUpcast;
```

Comment:
If you don't like this special case about smart pointers we can remove it. The presets will just have to add an `explicitCast` info on `std::shared_ptr<X>` in addition to the info on `X`.

Reply:
In fact `explicitUpcast` should really target a Java name, and not a C++ name. So what can be done is:
* when an info `explicitUpcast` is encountered for a C++ name, we register its associated Java name  in some `Set`.
* we set `type.explicitUpcast` to `true` when the java Name is in the set (or simply get rid of `type.explicitUpcast` and check `set.contains(type.javaName)` instead of checking`type.explicitUpcast`).
Thus no need for a specific code about smart Pointer like above. The fact that C++ `M` and `shared_ptr<M>` are associated with the same Java name is enough.

What do you think about this ?

Reply:
Looks good! Now the only place left where strings related to shared_ptr are used is in explicitCast(), so if you could replace that by the actual annotations on the constructor, it should work pretty much like I'm imagining it.

Reply:
I changed also this part to just check for the presence of an annotation. I though that's what you suggested. Have you checked the new code ?

Reply:
Yes, that's alright, but instead of doing if(constructorIsAnnotated(t)) ... @SharedPtr ... do something like String a = constructorAnnotations(); if (a != null) ... a ...

Reply:
Ok. Done.